### DISCUSSION TITLE
Saving model weights and then re-loading using 'serialization.from_bytes'

### DISCUSSION THREAD

**Discussion by varunsingh88** [google/flax#2110] [Q&A]
I am trying to save the trained weights for my model to reload again later. I have a pickle object that I am exporting through the 'save_weights' function, and then reloading the saved weights again through the load_Saved_weights() function.  As per FLAX documentation, when reloading from serialized data, the same weight structure needs to be passed into 'serialization.from_bytes()' function. However, what I see is that regardless of the structure of weights passed into serialization.from_bytes() function, the previously trained weights from the pickle file are loaded nonetheless. I was curious to know if this behavior is normal since I was expecting an error to show up if the structure of my newly initialized weights don't match the structure of saved weights but this does not appear to be the case here. 

```py
def save_weights(trained_weights, filename):
    pickle.dump(trained_weights, open(filename, 'wb'))

def load_saved_weights(weights, filename):
    pkl_file = pickle.load(open(filename, 'rb'))
    trained_wts = serialization.from_bytes(weights, pkl_file)
    return trained_wts

### Creating serialized weights after model training
bytes_output = serialization.to_bytes(weights)
filename = 'trained_weights_model.sav'
save_weights(bytes_output, filename)

### Initializing random weights. My original model was initialized with features [4,4,2,1] and input shape : 1x2
seed = random.PRNGKey(0)
model = model_name(features=[4,3,2,1])
weights = model.init(seed, jnp.ones(shape=[1,5]))
saved_wts = load_saved_weights(weights, 'trained_weights_model.sav')
```

**Updated by marcvanzee@ (May 16, 2022)**: wrapped code in Python codeblock to make it easier to read.

**Comment by cgarciae** [comment]
The `target` argument to `from_bytes` (`weights` in this example) currently seem to do 2 things:

* Raise errors if keys that are not present in the loaded structure are requested by `target`.
* Filters any additional key from the loaded structure not present in `target`.

However, it doesn't check for shape differences between the loaded structure and `target`, not sure if this is done intentionally or its a missing feature.  I've created #2436 to track this.
